### PR TITLE
Fix Alloca DataLayout for GPU compilation

### PR DESCRIPTION
**Pull request by collinwarner**
This PR fixes issues related to usage of the AMDGPU LLVM target by GPUCompiler. The AMDGPU target places stack allocations in LLVM addrspace 5, but many of our optimization passes assume that stack allocations are in addrspace 0 (which works for basically all other targets). This PR fixes that by querying the correct stack addrspace from the module's datalayout.

### CODE REVIEW

**Comment by maleadt** [comment]
The IRBuilder's `CreateAlloca` also queries the preferred alignment from the DL, is this something we may want to do too?

File: src/llvm-late-gc-lowering.cpp [hunk 1]
```diff
@@ -2247,8 +2246,9 @@ bool LateLowerGCFrame::CleanupIR(Function &F, State *S, bool *CFGModified) {
     PointerType *T_pprjlvalue = nullptr;
     AllocaInst *Frame = nullptr;
     if (T_prjlvalue) {
+       // assert(F.getParent()->getDataLayout().getAllocaAddrSpace() != 5 && "address space IS equal to 5");
```
Comment (jpsamaroo) [inline src/llvm-late-gc-lowering.cpp:2249]:
```suggestion
```

File: src/llvm-propagate-addrspaces.cpp [hunk 2]
```diff
@@ -165,15 +166,14 @@ Value *PropagateJuliaAddrspacesVisitor::LiftPointer(Value *V, Instruction *Inser
             Instruction *InstV = cast<Instruction>(V);
             Instruction *NewV = InstV->clone();
             ToInsert.push_back(std::make_pair(NewV, InstV));
-            Type *NewRetTy = PointerType::getWithSamePointeeType(cast<PointerType>(InstV->getType()), AddressSpace::Generic);
+            Type *NewRetTy = PointerType::getWithSamePointeeType(cast<PointerType>(InstV->getType()), currentAddressSpace);
```
Comment (jpsamaroo) [inline src/llvm-propagate-addrspaces.cpp:169]:
```suggestion
            Type *NewRetTy = PointerType::getWithSamePointeeType(cast<PointerType>(InstV->getType()), allocaAddressSpace);
```

File: src/llvm-propagate-addrspaces.cpp [hunk 3]
```diff
@@ -165,15 +166,14 @@ Value *PropagateJuliaAddrspacesVisitor::LiftPointer(Value *V, Instruction *Inser
...
             NewV->mutateType(NewRetTy);
             LiftingMap[InstV] = NewV;
             ToRevisit.push_back(NewV);
         }
     }
-
     auto CollapseCastsAndLift = [&](Value *CurrentV, Instruction *InsertPt) -> Value * {
-        PointerType *TargetType = PointerType::getWithSamePointeeType(cast<PointerType>(CurrentV->getType()), AddressSpace::Generic);
+        PointerType *TargetType = PointerType::getWithSamePointeeType(cast<PointerType>(CurrentV->getType()), currentAddressSpace);
```
Comment (jpsamaroo) [inline src/llvm-propagate-addrspaces.cpp:176]:
```suggestion
        PointerType *TargetType = PointerType::getWithSamePointeeType(cast<PointerType>(CurrentV->getType()), allocaAddressSpace);
```

**Comment by vtjnash** [comment]
This seems good to me.

For alignment, we are slowly moving away from LLVM guesses in favor of our own computations (since we have that from our own layout)

**Pull request closed without merging**